File: BOJ/14488.cpp
#include <cstdio>
#include <algorithm>

using namespace std;

pair<double, double> x[50001];
pair<double, double> sol;
int n;
double t;
long long a[50001];
long long v[50001];

pair<double, double> join(pair<double, double> z, pair<double, double> y)
{
	pair<double, double> ret;
	if (z.second > y.second)
	{
		ret.second = y.second;
		if (z.first > y.second) return pair<double, double>(-1.0, -1.0);
	}
	else
	{
		ret.second = z.second;
		if (y.first > z.second) return pair<double, double>(-1.0, -1.0);
	}
	ret.first = max(z.first, y.first);
	return ret;
}

int check()
{
	for (int i = 1; i < n; i++)
	{
		pair<double, double> tmp = join(x[i], x[i - 1]);
		sol = join(tmp, sol);
		if (sol.first < 0.0) return 0;
	}
	return 1;
}

int main()
{
	scanf("%d %lf", &n, &t);
	for (int i = 0; i < n; i++) scanf("%lld", a + i);
	for (int i = 0; i < n; i++) scanf("%lld", v + i);
	for (int i = 0; i < n; i++)
	{
		x[i].first = (double)a[i] - (double)v[i] * t;
		x[i].second = (double)a[i] + (double)v[i] * t;
		if (x[i].first < 1.0) x[i].first = 1.0;
		if (x[i].second > 1000000000.0) x[i].second = 1000000000.0;
	}
	sol = x[0];
	printf("%d", check());
	return 0;
}

File: BOJ/14488.py
def join(xx, yy):
    ret = [0, 0]
    if xx[1] > yy[1]:
        ret[1] = yy[1]
        if xx[0] > yy[1]:
            return [-1, -1]
    else:
        ret[1] = xx[1]
        if yy[0] > xx[1]:
            return [-1, -1]
    ret[0] = max(xx[0], yy[0])
    return ret
def check(kk, sol):
    for i in range(1, n):
        tmp = join(kk[i], kk[i-1])
        sol = join(tmp, sol)
        if sol[0] < 0:
            return 0
    return 1

x = input().split()
n = int(x[0])
cc = x[1]
t = 0
flag = 0
count = 0
for c in cc:
    if flag:
        count += 1
    if c == '.':
        flag = 1
    else:
        t *= 10
        t += int(c)
t *= 10**(4-count)
    
a = [int(x)*10000 for x in input().split()]
v = [int(x) for x in input().split()]
kk = []

for i in range(0, n):
    kk.append([a[i] - v[i] * t, a[i] + v[i] * t])
    if kk[i][0] < 10000:
        kk[i][0] = 10000
    if kk[i][1] > 10000000000000:
        kk[i][1] = 10000000000000

sol = kk[0]
print(check(kk, sol))